docs/design/2018-12-10-plugin-framework.md

```diff
@@ -0,0 +1,237 @@
+- package tool also add some build time and misc info into Manifest info
+
+Package tools add an abstract layer over manifest, so we can change manifest easier in future if needed. 
+
+#### Plugin Point
+
+In TiDB code, we can add new plugin point in everywhere and:
+
+- call `plugin.GetByKind` or `plugin.Get` to find matched plugins
+- call `plugin.Declare[Kind]Manifest` to cast Manifest to a special kind
+- call extension point method for special manifest
+
+we can see a simple example in `clientConn#Run` and `conn_ip_example` plugin implement. 
+
+Add new plugin point need to modify TiDB code and pass the required context and parameters.
+
+#### Configuration
+
+Every plugin has its own configurations, TiDB plugin use system variable to handle configuration management requirement.
+
+In `manifest.toml`, we can use `sysVar` field to provide plugin's variable name and its default value. Plugin's system variable will be registered as TiDB system variable, so the user can read/modify variable just like normal system variable.
+
+Plugin's variable name must use plugin name as the prefix. at last, the plugin cannot be reloaded if we change plugin's sysVar(include default-value, add or remove variable)
+
+We implement it by add plugin variable into `variable.SysVars` before `bootstrap`, so later `doDMLWorker` will handle them just as normal sysVars, and change `loadCommonGlobalVarsSQL` to load them. (cannot unload plugin and cannot modify sysVar during reload make this implement easier) 
+
+#### Dependency
+
+Go's plugin mechanism will check all dependency package hash to ensure link time and run time use some version([see code](https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/runtime/plugin.go#L52)), so we no longer need to take care compile package dependency.
+
+but for the real world, there is maybe logic dependency between plugin. for example, some guy writes an authorization plugin but it relies on vault plugin and only works if vault enables but isn't directly rely on vault plugin's source code.
+
+In `manifest.toml`, we can use `requireVersion` to declare A plugin require B plugin in X version, then plugin runtime will check it during load or reload phase.
+
+### Reload
+
+Go plugin doesn't support unload a plugin, but it can not stop us to load multiple version plugin into host program and framework ensure last reload one will be active, and others aren't unloaded but disable.
+
+So, we can reload plugin with a different version that is packaged by `pluginpkg` to modify plugin's implement logic, although we can not change the plugin's meta info(e.g. sysVars) now, I think it's still useful.
+
+#### Management
+
+For add a plugin to TiDB, we need:
+
+- add `-plugin-dir` as start argument to specify the folder contains plugins, e.g. '-plugin-dir=/data/deploy/tidb/plugin'
+- add `-plugin-load` as start argument to specify the plugin id(name "-" version) that need be load, e.g. '-plugin-load=conn_limit-1'
```

Comment:
If no `-plugin-load` specified, does MySQL load all plugins in `-plugin-dir`?

Reply:
Yes, it doesn't. normally mysql default installed with plugin like version_token.so and so on but they are not load until `-plugin-load` in config or `INSTALL` command